### PR TITLE
*: Convert CaseExpr to ScalarFunction and clean up

### PR DESCRIPTION
**Pull request by zimulala**

### CODE REVIEW

File: plan/expression_rewriter.go [hunk 1]
```diff
@@ -488,6 +498,17 @@ func (er *expressionRewriter) inToScalarFunc(v *ast.PatternInExpr) {
 	er.ctxStack = append(er.ctxStack, function)
 }
 
+func (er *expressionRewriter) caseToScalarFunc(v *ast.CaseExpr) {
+	stkLen := len(er.ctxStack)
+	argsLen := 2 * len(v.WhenClauses)
```
Comment (shenli) [inline plan/expression_rewriter.go]: Add some comments for the data layout.

**Comment by zimulala** [comment]
PTAL @shenli @hanfei19910905 @coocood

**Comment by coocood** [comment]
LGTM

File: evaluator/builtin_other.go [hunk 2]
```diff
@@ -83,6 +83,30 @@ func builtinOrOr(args []types.Datum, _ context.Context) (d types.Datum, err erro
...
+			return d, errors.Trace(err1)
+		}
+		if b == 1 {
+			d = args[i+1]
+			return
+		}
+	}
+	// when clause(condition, result) -> args[i], args[i+1]; (i >= 0 && i+1 < l-1)
+	// else clause -> args[l-1]
+	// If case clasue has else clause, l%2 == 1.
```
Comment (hanfei19910905) [inline evaluator/builtin_other.go]: clause not clasue

File: plan/expression_rewriter.go [hunk 3]
```diff
@@ -488,6 +498,45 @@ func (er *expressionRewriter) inToScalarFunc(v *ast.PatternInExpr) {
...
+	var args []expression.Expression
+	if v.Value != nil {
+		// args:  eq scalar func(args: value, condition1), result1,
+		//        eq scalar func(args: value, condition2), result2,
+		//        ...
+		//        else clasue
+		value := er.ctxStack[stkLen-argsLen-1]
+		args = make([]expression.Expression, 0, argsLen)
+		for i := stkLen - argsLen; i < stkLen-1; i += 2 {
+			arg := expression.NewFunction(ast.EQ, value.GetType(), value, er.ctxStack[i])
```
Comment (hanfei19910905) [inline plan/expression_rewriter.go]: the type of EQ is mysql.TypeTiny.

File: plan/expression_rewriter.go [hunk 4]
```diff
@@ -61,7 +61,7 @@ func getRowArg(e expression.Expression, idx int) expression.Expression {
 	}
 	c, _ := e.(*expression.Constant)
 	d := c.Value.GetRow()[idx]
-	return &expression.Constant{Value: d, RetType: types.NewFieldType(d.Kind())}
+	return datumToConstant(d, d.Kind())
```
Comment (hanfei19910905) [inline plan/expression_rewriter.go]: d.Kind() should be changed to c.GetType()

**Comment by zimulala** [comment]
PTAL @hanfei19910905

**Comment by hanfei19910905** [comment]
LGTM

**Pull request merged**